### DISCUSSION TITLE
Triton Intersection - Models not loaded

### DISCUSSION THREAD

**Discussion by marcoma9023** [pelicanmapping/osgearth#1909] [Support Q&A]
Hi,

I'm integrating Triton into OSGEarth and in general everything is fine.
But I'm having the following problem: when the Triton callback passes it calls the addIntersections method on the TritonLayerù+

`osg::ref_ptr<osgEarth::Triton::TritonIntersections> isect = new osgEarth::Triton::TritonIntersections;
_tritonLayer->addIntersections(isect);`

first thing i notice that, as time goes by, the ship models (on which my application is based) are not loaded, but i see as a LOD. In fact, when I approach the model, it is rendered. In the following video it is shown ..

https://user-images.githubusercontent.com/87537804/146733204-9ca6d16c-bbf9-46b0-aac2-53b94a227a5d.mp4

Let's say this effect can be tolerable.

The real problem is that sometimes, after the application runs for a while, templates don't load anymore.

Seeing the OSGEarth native code related to the intersections I noticed that in the TritonLayer there is a part that generates a kind of LOD for the intersections

`//INTERSECTIONS BETWEEN WAVES AND SHIP
			else if (nv.getVisitorType() == nv.CULL_VISITOR && _drawable && _TRITON.valid() && _TRITON->getOcean())
			{
				
				// Update any intersections.
				// For now this is running in the CULL traversal, which is not ideal.
				// However the Triton Ocean::GetHeight method requires a pointer to the Triton "camera"
				// and under our framework this is only available in CULL or DRAW.
				// Running the intersection in eithe CULL or DRAW will result in a frame
				// incoherency w.r.t the Triton update() call that updates the ocean state.
				::Triton::Ocean* ocean = _TRITON->getOcean();
				
				// Find the TritonCam associated with this osg Cam
				osgUtil::CullVisitor* cv = dynamic_cast<osgUtil::CullVisitor*>(&nv);
				::Triton::Camera* tritonCam = static_cast<TritonDrawable*>(_drawable)->getTritonCam(cv->getCurrentCamera());
				
				osg::Vec3d eye = osg::Vec3d(0, 0, 0) * cv->getCurrentCamera()->getInverseViewMatrix();

				for (std::vector<osg::ref_ptr<TritonIntersections> >::iterator i = _isect.begin();
					i != _isect.end();
					++i)
				{
					TritonIntersections* ir = i->get();

					// allocate enough space for the output:
					ir->_input.resize(ir->_input.size());
					ir->_normals.resize(ir->_input.size());

					osg::Matrix local2world;
					ir->_anchor.createLocalToWorld(local2world);

					// Make sure it's in range so as not to waste cycles:
					osg::Vec3d anchor = osg::Vec3d(0, 0, 0) * local2world;
					double m = ir->getMaxRange().as(osgEarth::Units::METERS);
					
					if ((eye - anchor).length2() > (m*m))
					{
						continue;
					}

					osg::Matrix world2local;
					world2local.invert(local2world);

					for (unsigned i = 0; i < ir->_input.size(); ++i)
					{
						const osg::Vec3d& local = ir->_input[i];

						// transform the ray to world coordinates
						osg::Vec3d start = local * local2world;
						osg::Vec3d dir = osg::Matrix::transform3x3(local2world, osg::Vec3d(0, 0, 1));

						// intersect the ocean
						float& out_height = ir->_heights[i];
						::Triton::Vector3 out_normalT;

						bool ok = ocean->GetHeight(
							::Triton::Vector3(start.x(), start.y(), start.z()),
							::Triton::Vector3(dir.x(), dir.y(), dir.z()),
							out_height,
							out_normalT,
							true,           // visualCorrelation
							true,           // includeWakes
							true,           // highResolution
							true,           // threadSafe,
							0L,             // intersectionPoint,
							true,           // autoFlip
							tritonCam);

						if (ok)
						{
							// populate the output data in local coordinates
							osg::Vec3d& normal = ir->_normals[i];
							normal.set(out_normalT.x, out_normalT.y, out_normalT.z);
							normal = osg::Matrix::transform3x3(normal, world2local);
						}
						else
						{
							// todo...what?
							OE_WARN << "GetHeight returned false dude" << std::endl;
						}
					}
				}
			}`

in particular the part 

`if ((eye - anchor).length2() > (m*m))
					{
						continue;
					}`

is it possible that sometimes, if this fails, the model is not loaded?

Thank you

**Comment by gwaldron** [comment]
This code is not involved in loading the model. All it does it find the height of the ocean at each point and return that height (and normal vector) in the intersections structure.

The "max range" just allows the code to skip intersections are all too far away; this is just a performance optimization. You can set the max range to FLT_MAX to disable this.

But again, it has nothing to do with loading the model.
